File: server/claude.ts
import fs from "node:fs";
import os from "node:os";
import path from "node:path";
import { LIVE_MS } from "../lib/config.ts";
import { peekJsonl, tryParseJson } from "../lib/node.ts";
import { visitFromClaudeRecord } from "../lib/extract.ts";
import type { LineParse, SessionHint, SessionRow } from "../lib/types.ts";
import {
  commandGroup,
  ensureHookList,
  hasPlexusLauncher,
  loadHooksFile,
  migrateLauncherCommands,
  saveHooksFile,
} from "./hooks.ts";

export function claudeHome(): string {
  return process.env.CLAUDE_CONFIG_DIR || path.join(os.homedir(), ".claude");
}

function titleFromRecords(records: Record<string, unknown>[], cwd: string | null): string {
  for (const record of records) {
    if (record?.type === "custom-title" && typeof record.title === "string" && record.title.trim()) {
      return record.title.trim();
    }
    if (record?.type === "summary" && typeof record.summary === "string" && record.summary.trim()) {
      return record.summary.trim();
    }
  }
  return path.basename(cwd || "") || "claude";
}

function cwdFromRecords(records: Record<string, unknown>[]): string | null {
  for (const record of records) {
    if (typeof record?.cwd === "string" && record.cwd) return record.cwd;
  }
  return null;
}

export function parseClaudeLine(line: string, session: SessionHint): LineParse {
  const record = tryParseJson<Record<string, unknown>>(line);
  if (!record) return { visits: [], activity: null, prompt: false };
  const type = record.type;
  const message = record.message && typeof record.message === "object" ? (record.message as Record<string, unknown>) : null;
  const content = message?.content;
  const isToolResult =
    type === "user" &&
    Array.isArray(content) &&
    content.some((block) => block && typeof block === "object" && (block as Record<string, unknown>).type === "tool_result");
  const prompt = type === "user" && !isToolResult;
  let activity: LineParse["activity"] = null;
  if (type === "assistant" || prompt) activity = "busy";
  if (record.hook_event_name === "Stop" || record.type === "stop") activity = "idle";
  const visits = visitFromClaudeRecord(record, session);
  return { visits, activity, prompt };
}

export function readClaudeSessions(home = claudeHome(), now = Date.now()): SessionRow[] {
  const root = path.join(home, "projects");
  if (!fs.existsSync(root)) return [];
  let projects: fs.Dirent[];
  try {
    projects = fs.readdirSync(root, { withFileTypes: true });
  } catch {
    return [];
  }
  const out: SessionRow[] = [];
  for (const project of projects) {
    if (!project.isDirectory()) continue;
    const dir = path.join(root, project.name);
    let files: string[];
    try {
      files = fs.readdirSync(dir);
    } catch {
      continue;
    }
    for (const name of files) {
      if (!name.endsWith(".jsonl")) continue;
      const file = path.join(dir, name);
      let st: fs.Stats;
      try {
        st = fs.statSync(file);
      } catch {
        continue;
      }
      if (now - st.mtimeMs > LIVE_MS) continue;
      const records = peekJsonl(file);
      const cwd = cwdFromRecords(records);
      if (!cwd) continue;
      const id = name.replace(/\.jsonl$/, "");
      out.push({
        session_id: `claude:${id}`,
        nativeId: id,
        pid: 0,
        cwd,
        title: titleFromRecords(records, cwd),
        agent: "claude",
        provider: "claude",
        updates: file,
        mtime: st.mtimeMs,
        live: true,
      });
    }
  }
  out.sort((a, b) => (b.mtime || 0) - (a.mtime || 0));
  return out;
}

export function installClaudeHooks({
  nodePath,
  launcher,
  hookBin,
  url,
}: {
  nodePath: string;
  launcher: string;
  hookBin: string;
  url: string;
}): void {
  const file = path.join(claudeHome(), "settings.json");
  const settings = loadHooksFile(file);
  migrateLauncherCommands(settings.hooks);
  const blob = JSON.stringify(settings.hooks);
  if (!hasPlexusLauncher(blob)) {
    ensureHookList(settings.hooks, "SessionStart").push(
      commandGroup(`${nodePath} "${launcher}" --ensure`, 8),
    );
  }
  if (!blob.includes("127.0.0.1:7733/hook") && !blob.includes("plexus-hook")) {
    ensureHookList(settings.hooks, "PreToolUse").push({
      matcher: "Read|Write|Edit|Glob|Grep|NotebookEdit",
      hooks: [{ type: "http", url, timeout: 2 }],
    });
    const sourceHook = commandGroup(`${nodePath} "${hookBin}" --source claude`, 2);
    ensureHookList(settings.hooks, "UserPromptSubmit").push(sourceHook);
    ensureHookList(settings.hooks, "Stop").push(sourceHook);
  }
  saveHooksFile(file, settings);
}

File: internal/claude/claude.go
package claude

import (
	"os"
	"path/filepath"
	"strings"
	"time"

	"github.com/ryusudol/plexus/internal/config"
	"github.com/ryusudol/plexus/internal/extract"
	"github.com/ryusudol/plexus/internal/hooks"
	"github.com/ryusudol/plexus/internal/jsonx"
	"github.com/ryusudol/plexus/internal/paths"
	"github.com/ryusudol/plexus/internal/types"
)

func Home() string {
	if env := os.Getenv("CLAUDE_CONFIG_DIR"); env != "" {
		return env
	}
	return filepath.Join(paths.Home(), ".claude")
}

func titleFromRecords(records []map[string]any, cwd string) string {
	for _, rec := range records {
		if jsonx.Str(rec["type"]) == "custom-title" {
			if t := strings.TrimSpace(jsonx.Str(rec["title"])); t != "" {
				return t
			}
		}
		if jsonx.Str(rec["type"]) == "summary" {
			if t := strings.TrimSpace(jsonx.Str(rec["summary"])); t != "" {
				return t
			}
		}
	}
	base := filepath.Base(cwd)
	if base == "" {
		return "claude"
	}
	return base
}

func cwdFromRecords(records []map[string]any) string {
	for _, rec := range records {
		if s := jsonx.Str(rec["cwd"]); s != "" {
			return s
		}
	}
	return ""
}

func ParseLine(line string, session extract.SessionHint) extract.LineParse {
	record := jsonx.Parse(line)
	if record == nil {
		return extract.LineParse{}
	}
	rec := jsonx.AsMap(record)
	typ := jsonx.Str(rec["type"])
	message := jsonx.AsMap(rec["message"])
	content := any(nil)
	if message != nil {
		content = message["content"]
	}
	isToolResult := typ == "user" && jsonx.Slice(content) != nil
	if isToolResult {
		isToolResult = false
		for _, block := range jsonx.Slice(content) {
			if bm := jsonx.AsMap(block); bm != nil && jsonx.Str(bm["type"]) == "tool_result" {
				isToolResult = true
				break
			}
		}
	}
	prompt := typ == "user" && !isToolResult
	activity := ""
	if typ == "assistant" || prompt {
		activity = "busy"
	}
	if jsonx.Str(rec["hook_event_name"]) == "Stop" || jsonx.Str(rec["type"]) == "stop" {
		activity = "idle"
	}
	visits := extract.VisitFromClaudeRecord(record, session)
	if visits == nil {
		visits = []extract.ParsedVisit{}
	}
	return extract.LineParse{Visits: visits, Activity: activity, Prompt: prompt}
}

func ReadSessions(home string, now time.Time) []types.SessionRow {
	if home == "" {
		home = Home()
	}
	if now.IsZero() {
		now = time.Now()
	}
	root := filepath.Join(home, "projects")
	if _, err := os.Stat(root); err != nil {
		return nil
	}
	projects, err := os.ReadDir(root)
	if err != nil {
		return nil
	}
	out := []types.SessionRow{}
	for _, project := range projects {
		if !project.IsDir() {
			continue
		}
		dir := filepath.Join(root, project.Name())
		files, err := os.ReadDir(dir)
		if err != nil {
			continue
		}
		for _, name := range files {
			if !strings.HasSuffix(name.Name(), ".jsonl") {
				continue
			}
			file := filepath.Join(dir, name.Name())
			st, err := os.Stat(file)
			if err != nil {
				continue
			}
			if now.Sub(st.ModTime()) > config.Live {
				continue
			}
			records := paths.PeekJSONL(file, 40)
			cwd := cwdFromRecords(records)
			if cwd == "" {
				continue
			}
			id := strings.TrimSuffix(name.Name(), ".jsonl")
			out = append(out, types.SessionRow{
				SessionID: "claude:" + id,
				NativeID:  id,
				PID:       0,
				Cwd:       cwd,
				Title:     titleFromRecords(records, cwd),
				Agent:     "claude",
				Provider:  "claude",
				Updates:   file,
				Mtime:     st.ModTime().UnixMilli(),
				Live:      true,
			})
		}
	}
	return types.NewestByID(out)
}

func Install(bin string) error {
	file := filepath.Join(Home(), "settings.json")
	settings := hooks.LoadFile(file)
	hm := jsonx.AsMap(settings["hooks"])
	if hm == nil {
		hm = map[string]any{}
		settings["hooks"] = hm
	}
	hooks.MigrateLauncherCommands(hm, bin)
	hooks.MigratePlexusHTTP(hm, bin, "claude")
	blob, _ := jsonMarshal(hm)
	if !hooks.HasPlexusLauncher(string(blob)) {
		hooks.PushGroup(hm, "SessionStart", hooks.CommandGroup(quote(bin)+" --ensure", 8))
	}
	sourceHook := hooks.CommandGroup(quote(bin)+" --hook --source claude", 2)
	for _, event := range []string{"PreToolUse", "UserPromptSubmit", "Stop"} {
		hooks.EnsureEventHook(hm, event, sourceHook)
	}
	settings["hooks"] = hm
	return hooks.SaveFile(file, settings)
}

func quote(s string) string { return `"` + s + `"` }

func jsonMarshal(v any) ([]byte, error) {
	return marshal(v)
}
